src/main/java/seedu/address/logic/parser/LessonAddCommandParser.java

```diff
@@ -0,0 +1,104 @@
+import static seedu.address.commons.core.Messages.MESSAGE_INVALID_COMMAND_FORMAT;
+import static seedu.address.logic.parser.CliSyntax.PREFIX_DATE;
+import static seedu.address.logic.parser.CliSyntax.PREFIX_END_TIME;
+import static seedu.address.logic.parser.CliSyntax.PREFIX_HOMEWORK;
+import static seedu.address.logic.parser.CliSyntax.PREFIX_RECURRING;
+import static seedu.address.logic.parser.CliSyntax.PREFIX_START_TIME;
+import static seedu.address.logic.parser.CliSyntax.PREFIX_SUBJECT;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import seedu.address.commons.core.index.Index;
+import seedu.address.logic.commands.LessonAddCommand;
+import seedu.address.logic.parser.exceptions.ParseException;
+import seedu.address.model.lesson.Date;
+import seedu.address.model.lesson.Homework;
+import seedu.address.model.lesson.MakeUpLesson;
+import seedu.address.model.lesson.RecurringLesson;
+import seedu.address.model.lesson.Subject;
+import seedu.address.model.lesson.Time;
+
+/**
+ * Parses input arguments and creates a new LessonAddCommand object.
+ */
+public class LessonAddCommandParser {
+    /**
+     * Parses the given {@code String} of arguments in the context of the LessonAddCommand
+     * and returns a LessonAddCommand object for execution.
+     * @throws ParseException if the user input does not conform the expected format
+     */
+    public LessonAddCommand parse(String args) throws ParseException {
+        requireNonNull(args);
+        ArgumentMultimap argMultimap =
+                ArgumentTokenizer.tokenize(args, PREFIX_RECURRING, PREFIX_DATE, PREFIX_START_TIME,
+                        PREFIX_END_TIME, PREFIX_SUBJECT, PREFIX_HOMEWORK);
+
+        if (!arePrefixesPresent(argMultimap, PREFIX_DATE, PREFIX_START_TIME, PREFIX_END_TIME, PREFIX_SUBJECT)) {
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, LessonAddCommand.MESSAGE_USAGE));
+        }
+
+        Index index;
+
+        try {
+            index = ParserUtil.parseIndex(argMultimap.getPreamble());
+        } catch (ParseException pe) {
+            throw new ParseException(String.format(MESSAGE_INVALID_COMMAND_FORMAT, LessonAddCommand.MESSAGE_USAGE), pe);
+        }
+
+        Date date = ParserUtil.parseDate(argMultimap.getValue(PREFIX_DATE).get());
+        Time startTime = ParserUtil.parseTime(argMultimap.getValue(PREFIX_START_TIME).get());
+        Time endTime = ParserUtil.parseTime(argMultimap.getValue(PREFIX_END_TIME).get());
+        Subject subject = ParserUtil.parseSubject(argMultimap.getValue(PREFIX_SUBJECT).get());
+        Set<Homework> homework = parseHomeworkForEdit(argMultimap.getAllValues(PREFIX_HOMEWORK))
+                .orElse(new HashSet<>());
+
+        // Check if end time is earlier than start time
+        if (startTime.getLocalTime().compareTo(endTime.getLocalTime()) > 0) {
```

Comment:
Perhaps you can extract out the logic here into a `boolean isValidTime = (start, end) => start.getLocalTime().compareTo(end.getLocalTime()) > 0;`. The logic abit dense at this section ><